addon/content/messageTags.jsx

```diff
@@ -54,170 +45,156 @@ MessageTag.propTypes = {
+  const [primaryString, secondaryStrings = []] = strings;
+  const primaryTooltip = <div>{primaryString}</div>;
+  const secondaryTooltip = secondaryStrings.length ? (
+    <React.Fragment>
+      <hr />
+      {secondaryStrings.map((s, i) => (
+        <div key={i}>{s}</div>
+      ))}
+      <div />
+    </React.Fragment>
+  ) : null;
+
+  return (
+    <span>
+      {primaryTooltip}
+      {secondaryTooltip}
+    </span>
+  );
+}
+DkimTooltip.propTypes = { strings: PropTypes.array.isRequired };
+
+function SpecialMessageTag(props) {
+  const {
+    icon,
+    name,
+    title = "",
+    tooltip = {},
+    onClick = null,
+    classNames,
+  } = props;
+
+  return (
+    <li
+      className={classNames + " special-tag" + (onClick ? " can-click" : "")}
+      title={title}
+      onClick={onClick}
+    >
+      <Icon path={icon} />
+      {name}
+      {tooltip.type === "dkim" && <DkimTooltip strings={tooltip.strings} />}
+    </li>
+  );
 }
 
 SpecialMessageTag.propTypes = {
-  canClick: PropTypes.bool.isRequired,
   classNames: PropTypes.string.isRequired,
-  dispatch: PropTypes.func.isRequired,
   icon: PropTypes.string.isRequired,
-  msgUri: PropTypes.string.isRequired,
   name: PropTypes.string.isRequired,
-  title: PropTypes.string.isRequired,
-  onClick: PropTypes.object.isRequired,
+  title: PropTypes.string,
+  onClick: PropTypes.func,
   tooltip: PropTypes.object.isRequired,
 };
 
-class SpecialMessageTags extends React.PureComponent {
-  constructor(props) {
-    super(props);
-    this.onClickInFolder = this.onClickInFolder.bind(this);
-  }
-
-  onClickInFolder() {
-    if (!this.props.canClickFolder) {
-      return;
-    }
-
-    this.props.dispatch({
-      type: "SWITCH_TO_FOLDER",
-      msgUri: this.props.msgUri,
-    });
-  }
-
-  render() {
-    return (
-      <ul className="tags special-tags">
-        {!!this.props.specialTags &&
-          !!this.props.specialTags.length &&
-          this.props.specialTags.map((tag, i) => {
-            return (
-              <SpecialMessageTag
-                canClick={tag.canClick}
-                classNames={tag.classNames}
-                dispatch={this.props.dispatch}
-                icon={tag.icon}
-                key={i}
-                msgUri={this.props.msgUri}
-                name={tag.name}
-                onClick={tag.onClick}
-                title={tag.title}
-                tooltip={tag.tooltip}
-              />
-            );
-          })}
-        {!!this.props.folderName && !this.props.inView && (
-          <li
-            className="in-folder"
-            onClick={this.onClickInFolder}
-            title={this.props.strings.get("jumpToFolder")}
-          >
-            {this.props.strings.get("inFolder", [this.props.folderName])}
-          </li>
-        )}
-      </ul>
+function SpecialMessageTags(props) {
+  const {
+    onTagClick,
+    onFolderClick = null,
+    specialTags,
+    strings,
+    inView,
+    folderName,
+  } = props;
```

Comment:
I just realised, how about we simplify these and do:

```
function SpecialMessageTags({
  onTagClick,
  onFolderClick = null,
  specialTags,
  strings,
  inView,
  folderName,
}) {
```
Then all the properties are documented directly in the function definition, and we don't need the separate breakdown.

Reply:
Sure, I can change this if you like. The reason I destructure the `props` instead of doing it directly in the function definition is that I've been bit in the past by error messages. `"props" is undefined` is a lot faster for me to recognize than `TypeError: (destructured parameter) is undefined`.